doc/barclamp.md

```diff
@@ -0,0 +1,1321 @@
+# rebase your working copy, so do not rebase your working copy. If you rebase
+# it (e.g. for final testing before a merge, please build a fresh Crowbar node
+# or restore it from a snapshot).
+function reposync_crowbar()
+  {
+  if [ -e /opt/dell/patchlevel ]; then
+    . /opt/dell/patchlevel
+    head=$(cd $BARCLAMP_FORK_PATH; git log --format='%H'| head -n 1)
+    if [ "$head" = "$commit" ]; then
+      echo "/opt/dell is already up to date. If this is not what you " 1>&2
+      echo "expected to see, you may have uncommitted changes" 1>&2
+      echo "in ${BARCLAMP_FORK_PATH}." 1>&2
+      return
+    fi
+  else
+    commit="${COMMIT}^"
+  fi
+
+  pushd $BARCLAMP_FORK_PATH > /dev/null
+  cp -a /opt/dell /opt/dell.orig
+  git diff $commit HEAD | patch -N --merge -p1 -d /opt/dell || return
+  rm -rf /opt/dell.orig
+  # Record current HEAD as patchlevel
+  echo "commit=$(git log --format='%H'| head -n 1)" > /opt/dell/patchlevel
+  popd > /dev/null
+
+  for barclamp in $BARCLAMPS
+    do
+    if [ ! -h /opt/dell/crowbar_framework/barclamps/${barclamp}.yml ]; then
+      ln -s /opt/dell/${barclamp}.yml /opt/dell/crowbar_framework/barclamps 2>&1 > /dev/null
+    fi
+    done
+  chmod 755 /opt/dell/bin/*
+
+  sync_crowbar
+  }
+
+# Helper function: updates Crowbar's runtime state after /opt/dell
+# has been modified.
+sync_crowbar()
+  {
+  barclamp_install.rb /opt/dell/crowbar_framework/barclamps &&
+  for barclamp in $BARCLAMPS
+    do
+    knife cookbook upload -o /opt/dell/chef/cookbooks/ $barclamp
+    done
+  systemctl restart crowbar
+  }
+
+```
+
+We will make heavy use of the `reposync_crowbar` command thus defined in the
+following sections. A short explanation of the three:
+
+`reposync_crowbar` will apply your whole fork against `/opt/dell` using
+`patch(1)` and sync Crowbar's runtime state with what is now in /opt/dell. You
+use this command to apply the current state of your fork to a freshly installed
+Crowbar node. This command may cause merge conflicts which you will have to
+resolve before proceeding. They usually happen due to a very out of date topic
+branch. While you can resolve them in /opt/dell you are better off rebasing
+your `crowbar-openstack` fork against upstream and applying the result to a
+fresh Crowbar node. Also, you will find a backup copy of /opt/dell in
+/opt/dell.orig if a merge conflict does occur.
+
+Using patches against `/opt/dell` may seem a bit over-the-top at first since
+one could simply copy files back and forth or use `barclamp_install` on your
+checkout, but it prevents a lot of breakage in cases where `/opt/dell` is out
+of sync with the `crowbar-openstack` repository's master branch or your own
+fork of `crowbar-openstack` is out of sync with upstream.
+
+In the former case copying files from your fork to `/opt/dell` or using
+`barclamp_install` may cause an inconsistent state and errors entirely
+unrelated to your change.
```

Comment:
That kind of misses the point.  You're right that the *other* RPMs can lag behind `master`, but I'm talking about the RPM which you're actually hacking on in this context, which is presumably `crowbar-openstack`.  My point was that you're generating a patch from the topic branch which is based off the (merge-)base of the topic branch, and then applying that patch against whatever revision of `crowbar-openstack` happens to be deployed in `/opt/dell`.  That's akin to a three-way merge, and is hence prone to not only merge conflicts (which are OK, because they're visible breakage which can be fixed) but also invisible breakage due to incompatibilities between the two base versions in the merge.

Having said all that, Dirk reminded me earlier today that in this particular context of creating a new barclamp, it is far less likely that this kind of breakage would occur, because all we're doing really is adding new files.  So it would only break if there was some incompatible API change in the barclamp framework, and that's fairly unlikely most of the time.

So in summary, I'm not too worried about the downsides of your approach *in this context*, although I'd recommend thinking carefully before using that in other "normal" hacking contexts.